Allowing load-balancing among multiple matching algorithm implementations

docularxu opened the pull request:
This patchset defined a new type of built-in provider with the name "loadbalance". The original idea was discussed in https://github.com/openssl/openssl/issues/19228. Here I'm proposing a draft, calling for your comments. Appreciate any type of feedbacks.

With this patchset, a user can specify what type of load balancing strategy to use when loading the 'loadbalance' provider. By default, it is round-robin. With this strategy (an integer number), the "loadbalance" provider creates a load_balancer libctx, which bases on child libctx, but without mirroring the “loadbalance” provider self into the child’s provider store. Inside the load_balancer libctx, a strategy related status is allocated and a scheduler function is bound.

The user can keep loading other providers which he wants to balance the work loads with into the parent libctx. As expected, these providers will be mirrored to the load_balancer libctx as child providers.

Suppose a dummy implementation of algorithm (eg. MD5) exists in both the “loadbalance” provider and the other providers, the user can now start a fetch with the property string “provider=loadbalance", and it will return the dummy implementation from the “loadbalance” provider.
`	md5 = EVP_MD_fetch(parent_libctx, "MD5", "provider=loadbalance");
`

Then at the init stage, i.e. digest init, the evp code will detect that this method comes from loadbalancer provider, so, it will go down one level, ie. check through the load_balancer libctx’s method store or cache, call previously bounded scheduler function to fetch a proper implementation from child providers, then set the new fetched method into context ctx.
`	EVP_DigestInit_ex(ctx, md5, NULL);
`

Moving onward, calculations will happen with the new method. 

When another thread calls the fetch/Init, with the help of the strategy specific scheduler function, it will be load-balanced to the next proper implementation, which may come from the same or not the same provider.

Some design notes, which I want to receive your comments :
1. I put the strategy specific scheduler functions in crypto/property/property.c, together with other method stores routines, and related struct types. I don’t want to export these structures and routines through OSSL_DISPATCH core_dispatch[].
2. Need to add more wrapper (dummy) algorithm implementations into the “loadbalance” provider. So we can load balance over them.
3. This feature is wrapped by #ifndef FIPS_MODULE. Because it relies on the child libctx.

This patchset include 7 patches:
Patch 1: Add load-balance strategy field into OSSL_LIB_CTX
Patch 2: Implement load-balance method fetch framework and add a round-robin strategy
Patch 3: Add a new field isloadbalancer in ossl_provider_st
Patch 4: Add a new load_balancer libctx type
Patch 5: Call load-balance fetch before normal method store and cache fetch
Patch 6: Digest init: fetch a real method from load_balancer libctx
Patch 7: Add a built-in provider loadbalance

Known issue:
1. crypto: function OSSL_LIB_CTX_new_load_balancer(3) undocumented
2. Tests, need to add more tests.

Usage Example:
```
{
    OSSL_LIB_CTX *parent_libctx = NULL;
    OSSL_PROVIDER *lbprov = NULL;
    OSSL_PROVIDER *deflt = NULL;

    lb_libctx = OSSL_LIB_CTX_new();

    /* load loadbalance provider, which by default enables round-robin strategy */
    lbprov = OSSL_PROVIDER_load(parent_libctx, "loadbalance");

    /* load other providers, and they will be mirrored into loadbalance provider's child libctx */
    deflt = OSSL_PROVIDER_load(parent_libctx, "default");

    {
	/*
	 * calculate MD5
	 */
	EVP_MD_CTX *ctx = NULL;
	EVP_MD *md5 = NULL;
	unsigned char *outdigest = NULL;

	/* specify property query sting: loadbalance */
	md5 = EVP_MD_fetch(parent_libctx, "MD5", "provider=loadbalance");

	/* Create a context for the digest operation */
	ctx = EVP_MD_CTX_new();
	/* Initialise the digest operation */
	EVP_DigestInit_ex(ctx, md5, NULL);
	EVP_DigestUpdate(ctx, msg, sizeof(msg));

	/* NOTE: cannot get_size from EVP_MD *md5, because the actual method has been 
	 * modified by EVP_DigestInit_ex(ctx)
	 */
	outdigest = OPENSSL_malloc(EVP_MD_get_size(EVP_MD_CTX_get0_md(ctx)));

	/* Now calculate the digest itself */
	EVP_DigestFinal_ex(ctx, outdigest, &len);
    }
    
    OSSL_PROVIDER_unload(lbprov);
    OSSL_PROVIDER_unload(deflt);
    OSSL_LIB_CTX_free(parent_libctx);
}

```

docularxu commented:
Updated. 
- Solved CLA requirement.
- Added one new test case for this feature.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

docularxu commented:
Based on the previous design, I implemented a free-bandwidth strategy. Method who has the biggest free-bandwidth wins the load_balancer_fetch().

    load-balance: implement free-bandwidth strategy
    
    This patch implements a free-bandwidth strategy for load-balance libctx.
    
    When free-bandwidth strategy is selected (see example config below),
    a separate free-bandwidth update thread will be created and it will poll
    each method provider periodically for free-bandwidth available. Based
    on the results, load_balancer_fetch() can always return the method who
    has the most free bandwidth.
    
    key design points:
    . added a thread to update free bandwidths
    . "f" is defined as the key word for querying free_bandwidth in provider's
      get_capabilities callback. Usage example:
        ossl_provider_get_capabilities(impl->provider, "f", NULL, &fbw_query);
    . allow the provider to return different free-bandwidth per method type, ie.
      <nid>. It means a provider can favor one algorithm than another, leaving
      the provider with more freedom on selecting which algorithm it will be
      allocated. Eg. a hardware accelerator may be more efficient in running
      AES than MD5. For this, it can return a 0 free-bandwidth on polling of MD5,
      and make load_balancer_fetch() return other provider's MD5 method. In
      contrast, it can return a big enough free-bandwidth number on AES and make
      load_balancer_fetch() select it over others.
    . update method's refcnt when adding to/removing from free_bandwidth's
      NID_best_impls data base.
    . use CRYPTO_atomic ops to set/access fetch_count
    
    Usage, an example config:
```
    [openssl_init]
    providers = providers
    
    [providers]
    default = default_conf
    loadbalance = loadbalance_conf
    
    [loadbalance_conf]
    activate = yes
    lb-strategy = 3
```

levitte commented on providers/build.info:
```diff
@@ -66,6 +66,17 @@ SOURCE[$LIBFIPS]=$LIBCOMMON
+#
+# load-balance provider stuff
+#
+# Currently, load-balance provider is built in, it means that libcrypto must
```
Considering that you have libcrypto share the provctx structure contents with this load-balance provider, it *must* remain built-in.

The general rule is that providers *must not share* the definition of opaque structures.  However, built-in providers may "cheat" in this regard, because it's guaranteed that the structures are exactly the same on both sides of the libcrypto\<-\>provider edge.

docularxu replied:
> The general rule is that providers _must not share_ the definition of opaque structures.

I understand this. The comments I put in build.info is not well thought-out. Sorry. I double checked the code of load-balance provider (lbprov.c). So far, it is safe to compile lbprov.c both as module and as static (built-in). I can provide this as a config option to the user in a later pull request, something similar to STATIC_LEGACY.

paulidale commented:
Instead of a capability, might it make sense to have the bandwidth be a param of the algorithm?

If this hurts performance too much, the relevant OSSL_PARAM could be queried once by the load balancer and the pointer to the data cached to avoid lots of calls to the get_param function.  Initially, probably don't bother.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

docularxu commented:
I verified this patch set on a 128-core Arm64 server with two types of MD5 providers: One is a vector multi-buffer [0] implementation "md5_mb" [1], the other is a hardware accelerator implementation "UADK" [2]. Result is promising. It gave the expected load-balancing bandwidth benefits, at a low overhead. My tests focused on two scenarios.
```
 [providers]
 loadbalance = loadbalance_conf
 libmd5mbprov = libmd5mbprov_conf
 uadk_provider = uadk_provider_conf
```
Scenario 1, from a CPU usage perspective, at a given CPU limit, leveraging both providers can significantly increase the bandwidth. Total bandwidth 
1. 300% CPU: 
> with "md5_mb" only: 1834 MB/s
> with load-balancing between both providers: "md5_mb" + "UADK": 3600 MB/s
2. 600% CPU:
> with "md5_mb" only: 3462 MB/s
> with load-balancing between both providers: "md5_mb" + "UADK": 7248 MB/s

Scenario 2, from a target MD5 bandwidth perspective, at a given bandwidth goal, the usage of CPU (PERCENT_CPU)  decrease as expected.
 1. 3600MB/s: 
> with "md5_mb" only: need 600% CPU
> with load-balancing between both providers: need only 330% CPU
2. 7000MB/s:
> with "md5_mb" only: need 1200% CPU
> with load-balancing between both providers: "md5_mb" + "UADK": need only 660% CPU

[0] multi-buffer: processing multiple independent data buffers simultaneously using SIMD instructions and registers
[1] md5_mb provider: [github link](https://github.com/docularxu/md5_mb_provider/tree/working_ossl_3.0_provider)
[2] UADK provider: [github link](https://github.com/Linaro/uadk_engine/pull/155)

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 278 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 309 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 340 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 371 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 402 days ago

nhorman commented:
Given that this or needs a rebase, and has been untouched for over a year, I'm closing it as inactive.  Please reopen if there is interest in picking it up again

t8m commented:
I believe this code should at least temporarily live in some repository outside of the primary openssl codebase. If it is deemed useful by sufficient number of users it could be merged.
